File: request-data.cpp
#include <unistd.h>
#include <cstdlib>
#include <iostream>
#include <ndn-cpp/face.hpp>

using namespace std;
using namespace ndn;
using namespace ndn::func_lib;

class Counter {
       public:
	Counter() { callbackCount_ = 0; }

	void onData(const ptr_lib::shared_ptr<const Interest>& interest,
		    const ptr_lib::shared_ptr<Data>& data) {
		++callbackCount_;
		cout << "Got data packet with name " << data->getName().toUri()
		     << endl;
		for (size_t i = 0; i < data->getContent().size(); ++i)
			cout << (*data->getContent())[i];
		cout << endl;
	}

	void onTimeout(const ptr_lib::shared_ptr<const Interest>& interest) {
		++callbackCount_;
		cout << "Time out for interest " << interest->getName().toUri()
		     << endl;
	}

	int callbackCount_;
};

int main(int argc, char** argv) {
	try {
		// Silence warning from Interest wire encode
		Interest::setDefaultCanBePrefix(true);

		Face face("155.98.38.244");

		// Counter holds data used by callback
		Counter counter;

		Name name("/ndn/external/test");
		cout << "Express name " << name.toUri() << endl;
		// Use bind to pass the counter object to the callbacks

		face.expressInterest(name,
				     bind(&Counter::onData, &counter, _1, _2),
				     bind(&Counter::onTimeout, &counter, _1));

		while (counter.callbackCount_ < 1) {
			face.processEvents();

			usleep(10000);
		}

	} catch (std::exception& e) {
		cout << "exception: " << e.what() << endl;
	}
	return 0;
}

File: request_data.py
import time
from pyndn import Name
from pyndn import Face
from pyndn import Interest
from pyndn.transport import UdpTransport
from pyndn.security import KeyChain


def dump(*list):
    """Prints all parameters"""

    result = ""
    for element in list:
        result += (element if type(element) is str else str(element)) + " "
    print(result)
    


class Counter():
    
    def __init__(self):
        self._callbackCount = 0

    
    def onData(self, interest, data):
        self._callbackCount += 1
        dump("Got data packet with name", data.getName().toUri())
        dump(data.getContent().toRawStr())


    def onTimeout(self, interest):
        self._callbackCount += 1
        dump("Time out for interest", interest.getName().toUri())


    def onNetworkNack(self, interest, networkNack):
        self._callbackCount += 1
        dump("Network nack for interest", interest.getName().toUri())


def main():

    # silence the warning from interest wire encode
    Interest.setDefaultCanBePrefix(True)

    # set up a face that connects to the remote forwarder
    ip_address = input("Enter an IP address to connect to: ")
    udp_connection_info = UdpTransport.ConnectionInfo(ip_address, 6363)
    udp_transport = UdpTransport()
    face = Face(udp_transport, udp_connection_info)

    counter = Counter()

    # try to fetch from provided name
    name_text = input("Enter a name to request content from: ")
    name = Name(name_text)
    dump("Express name", name.toUri())

    interest = Interest(name)
    interest.setMustBeFresh(False)
    face.expressInterest(interest, counter.onData, counter.onTimeout, counter.onNetworkNack)

    while counter._callbackCount < 1:
        face.processEvents()

        # don't use 100% of the CPU
        time.sleep(0.01)

    face.shutdown()


main()


